### PR TITLE
Implement codegen for queries

### PR DESCRIPTION
**Pull request by patrick91**
## Description

This is WIP 😊 This PR implements a feature similar to [GraphQL codegen](https://www.graphql-code-generator.com/) that allows to transform a GraphQL **query** into code, at the moment I'll be targeting only Python[1] code, but this can be expanded to other languages too (maybe via installable plugins).

[1] The python codegen will be a plugin. We'll be providing a generic interface that will make use of plugins to generate code.

Note: the code is kinda recreating `GraphQLType`, `GraphQLScalar` and so on. They are meant as tiny datastructures that will be used to generate code :)

TODO: 

There are some todo left in the code, on top of that we need:

- [x] Move codegen to a plugin (so we can have plugins for Python, TypeScript and so on)
- [x] Add a CLI interface (use rich?)
- [x] Proper errors
- [x] Improve internal interface for getting types (?)
- [x] Add plugin hooks to customise output 
- [x] Variables and arguments support
- [x] Support for returning multiple files 
- [x] Mark this as experimental in the CLI command

In other PRs:

- [ ] `pip install strawberry-graphql[cli]`
- [ ] Support for CLI options from the CLI plugins
- [ ] Plugin for future python codestyle (using `|, list`) could be enabled based on the version

/cc @peniqliotuv 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

### CODE REVIEW

**Comment by patrick91** [comment]
might be worth using this for tests: https://github.com/joseph-roitman/pytest-snapshot

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by botberry** [comment]
# Pre-release
:wave:

Pre-release **0.103.0-dev.1646681887** [dadda330cf0b057c1ee19625f012b188f86a0fef] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.103.0-dev.1646681887
```

File: strawberry/cli/commands/codegen.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        )
+    )
+
+    code_generator = QueryCodegen(schema_symbol, plugins=plugins)
+
+    with open(query) as f:
+        result = code_generator.codegen(f.read())
+
+    for file in result.files:
+        with open(f"{output_dir}/{file.path}", "w") as f:
```
Comment (estyxx) [inline strawberry/cli/commands/codegen.py:122]:
```suggestion
        with Path(f"{output_dir}/{file.path}").open("w") as f:
```
just some sugar :D

Comment (estyxx) [reply strawberry/cli/commands/codegen.py:122]:
```suggestion
        Path(f"{output_dir}/{file.path}").write_text(file.content)
```
oh even better!

Comment (patrick91) [reply strawberry/cli/commands/codegen.py]: updated :) I'm still working on the API for the CLI so it might change again :)

**Pull request merged**